src/main/kotlin/org/jitsi/videobridge/transport/ice/IceTransport.kt

```diff
@@ -0,0 +1,444 @@
+            // Once again, because the ICE Agent does not support adding
+            // candidates after the connectivity establishment has been started
+            // and because multiple transport-info JingleIQs may be used to send
+            // the whole set of transport candidates from the remote peer to the
+            // local peer, do not really start the connectivity establishment
+            // until we have at least one remote candidate per ICE Component.
+            if (iceComponent.remoteCandidateCount > 0) {
+                logger.info("Starting the agent with remote candidates.")
+                iceAgent.startConnectivityEstablishment()
+            }
+        } else if (iceStream.remoteUfragAndPasswordKnown()) {
+            // We don't have any remote candidates, but we already know the
+            // remote ufrag and password, so we can start ICE.
+            logger.info("Starting the Agent without remote candidates.")
+            iceAgent.startConnectivityEstablishment()
+        } else {
+            logger.cdebug { "Not starting ICE, no ufrag and pwd yet. ${transportPacketExtension.toXML()}" }
+        }
+    }
+
+    fun startReadingData() {
+        logger.cdebug { "Starting to read incoming data" }
+        val socket = iceComponent.socket
+        val receiveBuf = ByteBufferPool.getBuffer(1500)
+        val packet = DatagramPacket(receiveBuf, 0, receiveBuf.size)
+        var receivedTime: Instant
+
+        while (running.get()) {
+            try {
+                socket.receive(packet)
+                receivedTime = clock.instant()
+            } catch (e: SocketClosedException) {
+                logger.info("Socket closed, stopping reader")
+                break
+            } catch (e: IOException) {
+                logger.warn("Stopping reader", e)
+                break
+            }
+            packetStats.numPacketsReceived++
+            incomingDataHandler?.dataReceived(receiveBuf, 0, packet.length, receivedTime) ?: run {
+                logger.cdebug { "Data handler is null, dropping data" }
+                packetStats.numIncomingPacketsDroppedNoHandler++
+            }
+        }
+        logger.info("No longer running, stopped reading packets")
+    }
+
+    /**
+     * Send data out via this transport
+     *
+     * TODO: should we ensure that we're running before sending?
+     */
+    fun send(data: ByteArray, off: Int, length: Int) {
+        iceComponent.socket.send(DatagramPacket(data, off, length))
+        packetStats.numPacketsSent++
+    }
+
+    /**
+     * NOTE(brian): this method only exists because we want to use a Consumer<T> in
+     * SocketSenderNode in DtlsTransport.  When that gets ported to Kotlin it will be easier to
+     * make the T a function type (which can take multiple arguments), but until then we provide this.
+     */
+    fun send(packet: DatagramPacket) {
+        send(packet.data, packet.offset, packet.length)

```

Comment:
Why does `send(data: ByteArray, off: Int, length: Int)` exist? We take a `DatagramPacket`, break it down to its components, then create a new `DatagramPacket`.

Reply:
It's an annoying side effect (which I tried to describe in the comment) that will go away in the next phase (when `DtlsTransport` is redone).  `DtlsTransport` has function to set its "sender" (`setSender`) as we use a `Consumer` there.  `Consumer` only supports a single argument, so I had `DtlsTransport` take a `Consumer<DatagramPacket>`.  The other option would be to have `DtlsTransport` define a `OutgoingDataHandler` (and this is exactly what I did in the next phase of the refactor), but here in order to minimize the change I just added the shim in `IceTransport`, expecting that the next phase will follow this shortly (code is pretty much done there), so we won't be living with it for long.